File: 023-Merge-k-Sorted-Lists-priority-queue.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
from Queue import PriorityQueue
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """

        head = point = ListNode(0)
        q = PriorityQueue()
        for l in lists:
            if l:
                q.put((l.val, l))
        while not q.empty():
            val, node = q.get()
            point.next = ListNode(val)
            point = point.next
            node = node.next
            if node:
                q.put((node.val, node))
        return head.next

# In python, not python3
# Differences between python and python3:
# 1. python:  from Queue import PriorityQueue
#    python3: from queue import PriorityQueue
# 2. python:  q.put((l.val, l)) is correct
#    python3: TypeError: '<' not supported between instances of 'ListNode' and 'ListNode'
#    SOlution: add __lt__() function in class defination

# Runtime: 188ms
# Beats: 19.45%
# Memory beats: 7.03%

File: 023-Merge-k-Sorted-Lists-priority-queue.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto cmp = [](ListNode *&a, ListNode *&b)
        {
            return a->val > b->val;
        };
        priority_queue<ListNode*, vector<ListNode*>, decltype(cmp) > q(cmp);
        for(auto node: lists)
        {
            if(node) q.push(node);
        }
        
        ListNode *dummy = new ListNode(-1), *cur = dummy;
        while(!q.empty())
        {
            auto t = q.top();
            q.pop();
            cur -> next = t;
            cur = cur -> next;
            if(cur -> next) q.push(cur -> next);
        }
        return dummy -> next;
    }
};

// auto: 可以在声明变量的时候根据变量初始值的类型自动为此变量选择匹配的类型，
// 类似的关键字还有decltype。
// Runtime: 40ms
// Beats: 47.40%
// Memory beats: 99.63%
// Time complexity: O(nlogk)
// Space complexity: O(n) -> create a new list
//                   O(k) -> in-place method cost O(1)
//                   O(k) -> priority_queue